lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,778 @@
+	pub fn description(&self) -> &String {
+		&self.contents.description
+	}
+
+	/// Features for paying the invoice.
+	pub fn features(&self) -> Option<&OfferFeatures> {
+		self.contents.features.as_ref()
+	}
+
+	/// Duration since the Unix epoch when an invoice should no longer be requested.
+	///
+	/// If `None`, the offer does not expire.
+	pub fn absolute_expiry(&self) -> Option<Duration> {
+		self.contents.absolute_expiry
+	}
+
+	/// Whether the offer has expired.
+	#[cfg(feature = "std")]
+	pub fn is_expired(&self) -> bool {
+		match self.absolute_expiry() {
+			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
+				Ok(elapsed) => elapsed > seconds_from_epoch,
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	/// The issuer of the offer, possibly beginning with `user@domain` or `domain`.
+	pub fn issuer(&self) -> Option<&String> {
+		self.contents.issuer.as_ref()
+	}
+
+	/// Paths to the recipient originating from publicly reachable nodes.
+	pub fn paths(&self) -> Option<&Vec<BlindedPath>> {
+		self.contents.paths.as_ref()
+	}
+
+	/// The minimum quantity of items supported.
+	pub fn quantity_min(&self) -> u64 {
+		self.contents.quantity_min()
+	}
+
+	/// The maximum quantity of items supported.
+	pub fn quantity_max(&self) -> u64 {
+		self.contents.quantity_max()
+	}
+
+	/// The recipient's public key used to sign invoices.
+	pub fn node_id(&self) -> PublicKey {
+		self.contents.node_id.unwrap()
+	}
+
+	#[cfg(test)]
+	fn as_bytes(&self) -> &[u8] {
+		&self.bytes
+	}
+
+	#[cfg(test)]
+	fn to_bytes(&self) -> Vec<u8> {
+		let mut buffer = Vec::new();
+		self.contents.write(&mut buffer).unwrap();
+		buffer
+	}
+
+	#[cfg(test)]
+	fn as_tlv_stream(&self) -> reference::OfferTlvStream {
+		self.contents.as_tlv_stream()
+	}
+}
+
+impl OfferContents {
+	pub fn quantity_min(&self) -> u64 {
+		self.quantity_min.unwrap_or(1)
+	}
+
+	pub fn quantity_max(&self) -> u64 {
+		self.quantity_max.unwrap_or_else(||
+			self.quantity_min.map_or(1, |_| u64::max_value()))
+	}
+
+	fn as_tlv_stream(&self) -> reference::OfferTlvStream {
+		let (currency, amount) = match &self.amount {
+			None => (None, None),
+			Some(Amount::Bitcoin { amount_msats }) => (None, Some(amount_msats.into())),
+			Some(Amount::Currency { iso4217_code, amount }) => (
+				Some(iso4217_code), Some(amount.into())
+			),
+		};
+
+		reference::OfferTlvStream {
+			chains: self.chains.as_ref().map(Into::into),
+			metadata: self.metadata.as_ref().map(Into::into),
+			currency,
+			amount,
+			description: Some((&self.description).into()),
+			features: self.features.as_ref(),
+			absolute_expiry: self.absolute_expiry.map(|duration| duration.as_secs().into()),
+			paths: self.paths.as_ref().map(Into::into),
+			issuer: self.issuer.as_ref().map(Into::into),
+			quantity_min: self.quantity_min.map(Into::into),
+			quantity_max: self.quantity_max.map(Into::into),
+			node_id: self.node_id.as_ref(),
+		}
+	}
+}
+
+impl Writeable for OfferContents {
```

Comment:
For convenience/consistency with the rest of the API, might we want to implement `Writeable` for `Offer` as well?

Reply:
I'm a little reluctant to add this since `Offer`s are never directly serialized to bytes. The next PR adds parsing from and displaying as `bech32` strings. But maybe there is a use case that I'm not thinking of? Possibly open to it there is a strong reason. However, note that in the next PR `Offer` will implement `AsRef<[u8]>`, at very least.

Reply:
Oh, I figured since offers are designed to be long-lived users may want to persist them and reload them. Fine to punt it

Reply:
Yeah, that could be useful. Let's hold up until the next PR, though, which is a little smaller and adds the bech32 encoding/decoding at the `Offer` level.